File: leetcode-everyday/动态规划/基本动态规划-一维/198-打家劫舍/198-打家劫舍.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev = 0, next = nums[0], ans = nums[0];
        for(int i = 1; i < nums.size(); ++i) {
            ans = max(prev + nums[i], next);
            prev = next;
            next = ans;
        }
        return ans;
    }
};

File: leetcode-everyday/动态规划/基本动态规划-一维/198-打家劫舍/198-打家劫舍.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var rob = function(nums) {
  // const memo = new Array(nums.length+1).fill(0);;
  let ans = nums[0];
  // memo[1] = nums[0];
  let memo_next = nums[0];
  let memo_prev = 0;
  for(let i = 1; i < nums.length; i++) {
      // ans = Math.max(nums[i]+memo[i-1], memo[i]);
      ans = Math.max(nums[i] + memo_prev, memo_next);
      // memo[i+1] = Math.max(ans, memo[i+1]);
      memo_prev = memo_next;
      memo_next = ans;        
  }
  return ans;
};